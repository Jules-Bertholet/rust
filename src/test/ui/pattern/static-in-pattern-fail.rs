#![feature(static_in_pattern)]
#![feature(thread_local)]

struct NoPartialEqEq(i32, bool);

static mut ZERO: i32 = 0;
static ZERO_REF: &&f32 = &&0.0;
#[thread_local]
static TRUE: bool = true;
static FALSE: bool = false;
static WHEE: NoPartialEqEq = NoPartialEqEq(1, false);
#[link_name="IN"]
static OUT: i32 = 3;
extern "Rust" {
    static IN: i32;
}
static MAIN: fn() = crate::main;

fn main() {
     match 0 {
        ZERO => true,
        //~^ ERROR mutable statics cannot be referenced in patterns
        _ => false,
    };

     match (0, 1) {
        (ZERO, 1) => true,
         //~^ ERROR mutable statics cannot be referenced in patterns
        _ => false,
    };

     match 0 {
        IN => true,
        //~^ ERROR statics from `extern` blocks cannot be referenced in patterns
        _ => false,
    };

     match NoPartialEqEq(1, true) {
        WHEE => true,
        //~^ ERROR to use a constant or static of type `NoPartialEqEq` in a pattern, `NoPartialEqEq` must be annotated with `#[derive(PartialEq, Eq)]`
        _ => false,
    };

     match true {
        TRUE | FALSE => true,
        //~^ ERROR `#[thread_local]` statics cannot be referenced in patterns
        _ => false,
    };

     match crate::main as fn() {
        MAIN => true,
        //~^ ERROR function pointers cannot be used in patterns
        _ => false,
    };
}
