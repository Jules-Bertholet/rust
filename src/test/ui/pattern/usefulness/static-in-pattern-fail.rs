#![feature(static_in_pattern)]
#![feature(thread_local)]

struct NoPartialEqEq(i32, bool);

static mut ZERO: i32 = 0;
static ZERO_REF: &&f32 = &&0.0;
#[thread_local]
static TRUE: bool = true;
static FALSE: bool = false;
static WHEE: NoPartialEqEq = NoPartialEqEq(1, false);

fn main() {
    let val = match 0 {
        ZERO => true,
        //~^ ERROR mutable statics cannot be referenced in patterns
        _ => false,
    };

    let val = match (0, 1) {
        (ZERO, 1) => true,
         //~^ ERROR mutable statics cannot be referenced in patterns
        _ => false,
    };

    let val = match NoPartialEqEq(1, true) {
        WHEE => true,
        //~^ ERROR to use a constant or static of type `NoPartialEqEq` in a pattern, `NoPartialEqEq` must be annotated with `#[derive(PartialEq, Eq)]`
        _ => false,
    };

    let val = match true {
        TRUE | FALSE => true,
        //~^ ERROR `#[thread_local]` statics cannot be referenced in patterns
        _ => false,
    };
}
