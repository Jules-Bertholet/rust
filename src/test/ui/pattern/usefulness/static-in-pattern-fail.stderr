error[E0158]: mutable statics cannot be referenced in patterns
  --> $DIR/static-in-pattern-fail.rs:20:9
   |
LL |         ZERO => true,
   |         ^^^^

error[E0158]: mutable statics cannot be referenced in patterns
  --> $DIR/static-in-pattern-fail.rs:26:10
   |
LL |         (ZERO, 1) => true,
   |          ^^^^

error[E0158]: statics from `extern` blocks cannot be referenced in patterns
  --> $DIR/static-in-pattern-fail.rs:32:9
   |
LL |         IN => true,
   |         ^^

error: to use a constant or static of type `NoPartialEqEq` in a pattern, `NoPartialEqEq` must be annotated with `#[derive(PartialEq, Eq)]`
  --> $DIR/static-in-pattern-fail.rs:38:9
   |
LL |         WHEE => true,
   |         ^^^^

error[E0158]: `#[thread_local]` statics cannot be referenced in patterns
  --> $DIR/static-in-pattern-fail.rs:44:9
   |
LL |         TRUE | FALSE => true,
   |         ^^^^

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0158`.
